include/openmc/cell.h

```diff
@@ -130,21 +133,54 @@ class Cell
   //! \param on_surface The signed index of a surface that the coordinate is
   //!   known to be on.  This index takes precedence over surface sense
   //!   calculations.
+  virtual bool
+  contains(Position p, Direction u, int32_t on_surface) const = 0;
+
+  //! Find the oncoming boundary of this cell.
+  virtual std::pair<double, int32_t>
+  distance(Position p, Direction u, int32_t on_surface) const = 0;
+
+  //! Write all information needed to reconstruct the cell to an HDF5 group.
+  //! @param group_id An HDF5 group id.
+  virtual void to_hdf5(hid_t group_id) const = 0;
+
+};
+ 
+class CSGCell : public Cell
```

Comment:
Yes. Definitely. Thanks!

Reply:
This may be the wrong idea, but what I'm thinking now is that these attributes of the `Cell` class should have virtual getter/setter functions (maybe a few resize functions as well) which get called instead of accessing the class attributes directly. This would do a few things: 

  - Provide some protection for the attributes of these classes.

  - Maintain the use of a general `Cell` class throughout the geometry kernel (nothing new here I guess).

  - Allow some protection for improper use of the class depending on the simulation type. For example, access to anything other than the first cell instance in a `CADCell` should return an error or warning, likely indicating that is is being accessed by some CSG-specific part of the code when it shouldn't be. 

Thoughts?

Reply:
@paulromano still going back and forth on this in my head. Would it be better to just do the type conversion from `Cell` to `CSG/DAGCell` explicitly in the code?

Reply:
I was mostly thinking about members like `region` and `rpn` that are very CSG-specific. Those we could probably move into `CSGCell` without needing getters/setters. Were you thinking about other members?

Reply:
I agree with @pshriwise, that it'd be tough to move any of these members to `CSGCell` without a bunch of type-casting.  I think the really proper way to handle this issue is to have a `CSGGeometryKernel` that unconditionally keeps cells of the type `CSGCell` and a `DAGMCGeometryKernel` with `DAGCell`s, but that's a big enough change that I'm happy to leave it out of this PR.

Reply:
Are we ok with punting this to another PR after some discussion on the design?

Reply:
After looking at this this morning, there is only one place in the code which requires a type-cast to get to the `region_` attribute. This isn't too bad, but as @smharper said, probably not something to let become commonplace.